auth/certificate_authenticator.cc

```diff
@@ -0,0 +1,152 @@
+    return {};
+}
+
+auth::authentication_option_set auth::certificate_authenticator::alterable_options() const {
+    return {};
+}
+
+future<auth::authenticated_user> auth::certificate_authenticator::authenticate(const std::string& subject) const {
+    clogger.debug("Authenticating subject: {}", subject);
+
+    std::smatch m;
+    if (!std::regex_search(subject, m, _subject_expr)) {
+        throw exceptions::authentication_exception(format("Subject '{}' does not match query expression", subject));
+    }
+    auto username = m[1].str();
+    clogger.debug("Return username: {}", username);
+    return make_ready_future<authenticated_user>(username);
+}
+
+future<auth::authenticated_user> auth::certificate_authenticator::authenticate(const credentials_map& credentials) const {
+    auto i = credentials.find(subject_key);
+    if (i == credentials.end()) {
+        throw exceptions::authentication_exception(format("Required key '{}' is missing", subject_key));
+    }
+    return authenticate(i->second);
+}
+
+future<> auth::certificate_authenticator::create(std::string_view role_name, const authentication_options& options) const {
+    // TODO: should we keep track of roles/enforce existence? Role manager should deal with this...
+    co_return;
+}
+
+future<> auth::certificate_authenticator::alter(std::string_view role_name, const authentication_options& options) const {
+    co_return;
+}
+
+future<> auth::certificate_authenticator::drop(std::string_view role_name) const {
+    co_return;
+}
+
+future<auth::custom_options> auth::certificate_authenticator::query_custom_options(std::string_view) const {
+    co_return auth::custom_options{};
+}
+
+const auth::resource_set& auth::certificate_authenticator::protected_resources() const {
+    static const resource_set resources;
+    return resources;
+}
+
+::shared_ptr<auth::sasl_challenge> auth::certificate_authenticator::new_sasl_challenge(session_dn_func f) const {
+    if (!f) {
+        throw exceptions::authentication_exception("Certificate subject info required");
+    }
+    class cert_sasl_challenge : public sasl_challenge {
+        const certificate_authenticator& ca;
+        session_dn_func func;
+    public:
+        cert_sasl_challenge(const certificate_authenticator& ca_in, session_dn_func f)
+            : ca(ca_in)
+            , func(std::move(f))
+        {}
+        bytes evaluate_response(bytes_view client_response) {
+            return {};
+        }
+        bool is_complete() const {
+            return true;
+        }
+        future<authenticated_user> get_authenticated_user() const {
+            auto dninfo = co_await func();
+            if (!dninfo) {
+                throw exceptions::authentication_exception("No valid certificate found");
+            }
+            co_return co_await ca.authenticate(dninfo->subject);
+        }
+    };
+    return ::make_shared<cert_sasl_challenge>(*this, std::move(f));
```

Comment:
I don't completely understand the details here, but how is this related to sasl challenge/response? We aren't sending a challenge to the client, we're just evaluating already-provided information from TLS.

Why do the dance with pretending to be SASL while in practice we're not?

As a simplification I'd replace create_immediate_sasl_response() with try_authenticate_directly() or something. And transport/server.cc can do the extraction of the TLS information since that's what knows that we're indeed using TLS, while the various authenticators don't know if we're using TLS.

Maybe we don't even need to involve the authenticator. We're bypassing sasl so let's not pretend to be sasl.

Reply:
We need to involve the authenticator to handle resolved subjects (which are just mapping to roles). 
While it is true that cert DN has little to do with SASL, you can argue the session_dn_func argument is optional, and only provided iff sasl request is done via a certificate aware context (not even wire, just something involving certificates - remember that TLS is not same as sockets. And certificates are not same as TLS).

Reply:
But there is clearly a mismatch here between sasl and certificates.

sasl = negotiating proof of authenticity with the client, via a well-defined layered protocol on top of a less-defined rpc protocol (cql here)
certificates = client already proved their identity, we just need to translate to roles

Why even go through the authenticator? Just look up the name via role_manager. Authentication is already done.

Reply:
I guess that could make sense. Instead of `new_sasl_challenge`, simply add an extra call to the auth interface, something akin to the already added `create_immediate_sasl_response`, but call it something like `immediate_dn_auth`. Only downside it having to replicate some logic from `process_auth_response`, but in the end it might look cleaner.